src/transaction.ts

```diff
@@ -34,27 +34,56 @@ export class Transaction {
     throw new Error("Wallet was not set");
   }
 
+  /**
+   * Wait for the transaction to be mined or rejected.
+   * @returns The transaction status by asking the blockchain.
+   */
+  public async wait(): Promise<TransactionStatus> {
+    if (!!this.wallet.ethereum) {
+      await this.ethereumWait();
+      return this.ethereumStatus();
+    }
+    throw new Error("Wallet was not set");
+  }
+
   private async ethereumStatus(): Promise<TransactionStatus> {
     const wallet = this.ethereumWallet;
-    const receipt = await wallet.getTransactionReceipt(this.id);
-    if (!receipt) {
-      throw new Error(`Could not retrieve receipt for ${this.id} on Ethereum`);
-    }
-    if (receipt.status === undefined || receipt.status === 0) {
-      return TransactionStatus.Failed;
-    }
+
     const transaction = await wallet.getTransaction(this.id);
     if (!transaction) {
       throw new Error(
         `Could not retrieve transaction for ${this.id} on Ethereum`
       );
     }
     if (transaction.confirmations === 0) {
+      // If the transaction is not yet mined then its status is uncertain.
       return TransactionStatus.Pending;
     }
+
+    const receipt = await wallet.getTransactionReceipt(this.id);
+    if (!receipt) {
+      throw new Error(
+        `Could not retrieve receipt for ${this.id} on Ethereum despite ${transaction.confirmations} confirmations.`
+      );
+    }
+    if (receipt.status === undefined || receipt.status === 0) {
+      return TransactionStatus.Failed;
+    }
+
     return TransactionStatus.Confirmed;
   }
 
+  private async ethereumWait(): Promise<void> {
+    const wallet = this.ethereumWallet;
+
+    const transaction = await wallet.getTransaction(this.id);
+    try {
+      await transaction.wait(1);
+    } catch (e) {
+      // If it rejects it means it's mined and failed.
```

Comment:
❓ Are we 100% sure we want to swallow that exception? Shouldn't there be at least a log statement?

Reply:
The returned status will tell you if it is failed so it's not "swallowed".

Sorry I should have removed review request, redesigning with @yosriady's input.